### PR TITLE
Handle Client.Close on cloned clients

### LINKED ISSUE #886: independent lifecycles for cloned clients
## Expected Behavior

When using `NewClientFromExisting` from #881, the "parent" and "child" `Client`s have independent lifecycles from the point of view of client code, even though they share some resources. That is, I can call `Close` on only one or both of them, and resources will be cleaned up when appropriate.

This makes it easier to use in large programs since we don't have to essentially ban use of `Close`, we can let anyone who gets a cloned client treat it as if it were a fully-independent one.

## Actual Behavior

Calling `Close()` on either the parent or child will close the connection for both of them immediately.

## Steps to Reproduce the Problem

  1.
  1.
  1.

## Specifications

  - Version: unreleased post 1.16
  - Platform:

### PR DESCRIPTION
**Pull request by cretz**
## What was changed

Use a counter to make sure only the last `Close()` of a shared-connection set of clients applies.

## Checklist

1. Closes #886

### CODE REVIEW

File: internal/internal_workflow_client.go [hunk 1]
```diff
@@ -942,11 +948,25 @@ func (wc *WorkflowClient) ensureInitialized() error {
 
 // Close client and clean up underlying resources.
 func (wc *WorkflowClient) Close() {
-	if wc.conn == nil {
-		return
+	// If there's a set of unclosed clients, we have to decrement it and then
+	// set it to a new pointer of max to prevent decrementing on repeated Close
+	// calls to this client. If the count has not reached zero, this close call is
+	// ignored.
+	if wc.unclosedClients != nil {
```
Comment (dnr) [inline internal/internal_workflow_client.go:955]: this shouldn't ever be nil, right?

File: internal/internal_workflow_client.go [hunk 2]
```diff
@@ -942,11 +948,25 @@ func (wc *WorkflowClient) ensureInitialized() error {
...
+		wc.unclosedClients = &maxUnclosedClients
+		// If there are any remaining, do not close
+		if remainingUnclosedClients > 0 {
+			return
+		}
 	}
-	if err := wc.conn.Close(); err != nil {
-		wc.logger.Warn("unable to close connection", tagError, err)
+
+	if wc.conn != nil {
```
Comment (dnr) [inline internal/internal_workflow_client.go:966]: when is this nil?

File: internal/internal_workflow_client_test.go [hunk 3]
```diff
@@ -1402,3 +1404,45 @@ func serializeEvents(events []*historypb.HistoryEvent) *commonpb.DataBlob {
...
+	// Create primary client
+	server, err := startTestGRPCServer()
+	require.NoError(t, err)
+	defer server.Stop()
+	client, err := DialClient(ClientOptions{HostPort: server.addr})
+	require.NoError(t, err)
+	workflowClient := client.(*WorkflowClient)
+
+	// Confirm there is 1 unclosed client
+	require.EqualValues(t, 1, *workflowClient.unclosedClients)
```
[the remark this answers is not in the document]
Comment (cretz) [reply internal/internal_workflow_client_test.go:1418]: This is in a test case. I can trust my lack of race issues in my own test case.

File: internal/internal_workflow_client.go (hunk 1 above)
Comment (cretz) [reply internal/internal_workflow_client.go:955]: I cannot trust all places that create a client will set this. That includes tests and other code paths.

File: internal/internal_workflow_client.go (hunk 2 above)
Comment (cretz) [reply internal/internal_workflow_client.go:966]: Test cases w/ mocks. This logic has not changed, I just moved the condition from the top of the function to down here.

**Review by dnr — APPROVED** [review]
sounds good to me

**Comment by cretz** [comment]
This is reliably causing integration tests to go from 2 mins to 3. I am investigating...

File: internal/internal_workflow_client_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Sushisource) [reply internal/internal_workflow_client_test.go:1418]: I would do it just for consistency's sake. It's supposed to be an atomic int, it should be manipulated atomically even if Go lets you not do that. Trivial amount of extra typing.

**Review by Sushisource — APPROVED** [review]

File: internal/internal_workflow_client_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (cretz) [reply internal/internal_workflow_client_test.go:1418]: I also access the pointers directly in a non-atomic way to compare they point to the same spot. But I will make the change.

**Pull request merged**